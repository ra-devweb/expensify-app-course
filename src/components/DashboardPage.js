import React from 'react';

import ExpensesList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilters';

const DashboardPage = () => (
  <div>
    <ExpenseListFilter />
    <ExpensesList />
  </div>
);

export default DashboardPage;
