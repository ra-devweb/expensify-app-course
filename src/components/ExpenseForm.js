import React from 'react';
import { SingleDatePicker } from 'react-dates';
import moment from 'moment';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: '',
    };
  }

  onChangeDescription = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onChangeNote = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onChangeAmount = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onChangeDate = (createdAt) => {
    if (createdAt) this.setState(() => ({ createdAt }));
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmitForm = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount.',
      }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmitForm}>
          <div>
            <label htmlFor='description'>Description</label>
            <input
              type='text'
              id='description'
              placeholder='Description'
              value={this.state.description}
              onChange={this.onChangeDescription}
              autoFocus
            />
          </div>
          <div>
            <label htmlFor='amount'>Amount</label>
            <input
              id='amount'
              type='text'
              placeholder='Amount'
              value={this.state.amount}
              onChange={this.onChangeAmount}
            />
          </div>
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onChangeDate}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <div>
            <label htmlFor='note'>Note (Optional)</label>
            <textarea
              id='note'
              placeholder='Add a note for expense'
              value={this.state.note}
              onChange={this.onChangeNote}
            ></textarea>
          </div>
          <button>Add expense</button>
        </form>
      </div>
    );
  }
}
