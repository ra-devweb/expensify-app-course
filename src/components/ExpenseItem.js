import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseItem = ({ id, description, amount, createdAt }) => {
  return (
    <div>
      <Link to={`/edit/${id}`}>
        <h1>{description}</h1>
      </Link>
      <p>{amount}</p>
      <span>({createdAt})</span>
    </div>
  );
};

export default ExpenseItem;
