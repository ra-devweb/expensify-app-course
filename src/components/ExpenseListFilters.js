import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';

import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate,
} from '../actions/filters';

export class ExpenseListFilter extends React.Component {
  state = {
    focusedInput: null,
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (focusedInput) => {
    this.setState(() => ({
      focusedInput,
    }));
  };

  onTextChange = (e) => this.props.setTextFilter(e.target.value);

  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div className='filters'>
        <div className='container'>
          <div className='form text-center'>
            <div className='form__group'>
              <input
                className='form__input'
                placeholder='Search expenses'
                type='text'
                value={this.props.filters.text}
                onChange={this.onTextChange}
              />
            </div>
            <div className='form__group'>
              <select 
                className='form__select'
                value={this.props.filters.sortBy} onChange={this.onSortChange}
                >
                <option value='date'>Date</option>
                <option value='amount'>Amount</option>
              </select>
            </div>
            <div className='form__group'>
              <DateRangePicker
                startDate={this.props.filters.startDate}
                startDateId='startDateId'
                endDate={this.props.filters.endDate}
                endDateId='endDateId'
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.focusedInput}
                onFocusChange={this.onFocusChange}
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={() => false}
              />
            </div>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters,
});

const mapDispatchToProps = (dispatch) => ({
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);
