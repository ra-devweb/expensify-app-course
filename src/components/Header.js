import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogout } from '../actions/auth';
import { history } from '../routes/AppRouter';

export const Header = ({ logOut }) => (
  <header className='header'>
    <div className='container'>
      <nav className='navigation'>
        <ul className='navigation__list'>
          <li className='navigation__item'>

            { history.location.pathname === '/dashboard' ? '' : <span className='navigation__go-back' onClick={() => history.goBack()}>&#10229;</span>}

            &nbsp;
            
            <Link className='navigation__link' to='/dashboard'>
              <h1 className='navigation__logo'>Expensify</h1>
            </Link>
          </li>
        </ul>
        <button className='navigation__logout btn-small btn-small--back-h' onClick={logOut}>Logout</button>
      </nav>
    </div>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  logOut: () => dispatch(startLogout()),
});

export default connect(undefined, mapDispatchToProps)(Header);
