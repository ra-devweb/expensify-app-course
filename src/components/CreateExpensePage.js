import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export class CreateExpensePage extends React.Component {
  onSubmit = ({ description, amount, createdAt, note }) => {
    this.props.addExpense({ description, amount, createdAt, note });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addExpense: (expense) => dispatch(addExpense(expense)),
  };
};

export default connect(undefined, mapDispatchToProps)(CreateExpensePage);
