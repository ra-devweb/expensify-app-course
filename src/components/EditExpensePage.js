import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { updateExpense, deleteExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  updateExpense = (expense) => {
    this.props.updateExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  deleteExpense = () => {
    this.props.deleteExpense(this.props.expense.id);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.updateExpense}
        />
        <button onClick={this.deleteExpense}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      (expense) => expense.id === props.match.params.id
    ),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    updateExpense: (id, expense) => dispatch(updateExpense(id, expense)),
    deleteExpense: (id) => dispatch(deleteExpense({ id })),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
