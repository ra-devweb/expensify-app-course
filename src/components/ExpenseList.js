import React from 'react';
import { connect } from 'react-redux';

import selectExpenses from '../selectors/expenses';
import ExpenseItem from './ExpenseItem';

export const ExpenseList = (props) => (
  <div>
    {props.expenses.length === 0 ? (
      <p>No expense</p>
    ) : (
      props.expenses.map((expense) => {
        return <ExpenseItem key={expense.id} {...expense} />;
      })
    )}
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpenseList);
