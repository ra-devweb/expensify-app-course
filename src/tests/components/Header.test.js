import React from 'react';
import { shallow } from 'enzyme';

import { Header } from '../../components/Header';

test('Should render Header component correctly', () => {
  const wrapper = shallow(<Header logOut={() => {}} />);
  expect(wrapper).toMatchSnapshot();
});

test('Should startLogout on click button', () => {
  const logOut = jest.fn();
  const wrapper = shallow(<Header logOut={logOut} />);
  wrapper.find('button').simulate('click');
  expect(logOut).toHaveBeenCalled();
});
