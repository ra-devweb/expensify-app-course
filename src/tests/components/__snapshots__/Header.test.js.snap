// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Header component correctly 1`] = `
<header>
  <h1>
    Expensify
  </h1>
  <nav>
    <ul>
      <li>
        <NavLink
          activeClassName="is-active"
          exact={true}
          to="/"
        >
          Dashboard
        </NavLink>
      </li>
      <li>
        <NavLink
          activeClassName="is-active"
          to="/Create"
        >
          Create
        </NavLink>
      </li>
      <li>
        <NavLink
          activeClassName="is-active"
          to="/help"
        >
          Help
        </NavLink>
      </li>
    </ul>
  </nav>
</header>
`;
